samples/server/main.go

```diff
@@ -0,0 +1,127 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Command server runs a simple HTTP server with integrated StackDriver tracing
+// and health checks.
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"go.opencensus.io/trace"
+	"gocloud.dev/gcp"
+	"gocloud.dev/health"
+	"gocloud.dev/server"
+	"gocloud.dev/server/sdserver"
+)
+
+// GlobalMonitoredResource implements monitoredresource.Interface to provide a
+// basic global resource based on the project ID. If you're running this sample
+// on GCE or EC2, you may prefer to use monitoredresource.Autodetect() instead.
+type GlobalMonitoredResource struct {
+	projectId string
+}
+
+func (g GlobalMonitoredResource) MonitoredResource() (string, map[string]string) {
+	return "global", map[string]string{"project_id": g.projectId}
+}
+
+func helloHandler(w http.ResponseWriter, req *http.Request) {
+	fmt.Fprintf(w, "Hello\n")
+}
+
+func mainHandler(w http.ResponseWriter, req *http.Request) {
+	fmt.Fprintf(w, "Welcome to the home page!\n")
+}
+
+// customHealthCheck implements the health.Checker interface and reports the
+// server is healthy when the healthy field is set to true.
+type customHealthCheck struct {
+	mu      sync.RWMutex
+	healthy bool
+}
+
+func (h *customHealthCheck) CheckHealth() error {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+	if !h.healthy {
+		return errors.New("not ready yet!")
+	}
+	return nil
+}
+
+func main() {
+	addr := flag.String("listen", ":8080", "HTTP port to listen on")
+	doTrace := flag.Bool("trace", true, "Export traces to Stackdriver")
+	flag.Parse()
+
+	ctx := context.Background()
+	credentials, err := gcp.DefaultCredentials(ctx)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+	tokenSource := gcp.CredentialsTokenSource(credentials)
+	projectId, err := gcp.DefaultProjectID(credentials)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var exporter trace.Exporter
+	if *doTrace {
+		fmt.Println("Exporting traces to StackDriver")
+		mr := GlobalMonitoredResource{projectId: string(projectId)}
+		exporter, _, err = sdserver.NewExporter(projectId, tokenSource, mr)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/hello", helloHandler)
+	mux.HandleFunc("/", mainHandler)
+
+	// healthCheck will report the server is unhealthy for 10 seconds after
+	// startup, and as healthy henceforth. Check the /healthz/readiness
+	// HTTP path to see readiness.
+	healthCheck := new(customHealthCheck)
+	time.AfterFunc(10*time.Second, func() {
+		healthCheck.mu.Lock()
+		defer healthCheck.mu.Unlock()
+		healthCheck.healthy = true
+	})
+
+	options := &server.Options{
+		RequestLogger:         sdserver.NewRequestLogger(),
+		HealthChecks:          []health.Checker{healthCheck},
+		TraceExporter:         exporter,
+		DefaultSamplingPolicy: trace.AlwaysSample(),
```

Comment:
Probably want to add a note here that this not a good setting for a production server: a production server should use [`ProbabilitySampler`](https://godoc.org/go.opencensus.io/trace#ProbabilitySampler) with some reasonably low fraction.

(It is an incredibly useful setting for making sure your tracing exporter works, tho.)

Reply:
Done